src/LockTable.h

```diff
@@ -0,0 +1,187 @@
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef RAMCLOUD_LOCKTABLE_H
+#define RAMCLOUD_LOCKTABLE_H
+
+#include "Common.h"
+
+#include "Atomic.h"
+#include "Fence.h"
+
+namespace RAMCloud {
+
+/**
+ * The LockTable represents a collection a locks with unique IDs.  The locks can
+ * be acquired and released like any typical lock based our their lockId. The
+ * only restriction on the 64-bit lockId is that the value 0 can not be used.
+ *
+ * This class is used, for instance, to manage per %RAMCloud object write locks
+ * that protect against concurrent write transactions and write requests on the
+ * same object.  In this case, the lockId can be the address of the object in
+ * the %RAMCloud log.
+ *
+ * This class is thread-safe.
+ *
+ * \section impl Implementation Details
+ *
+ * Locks represented in the LockTable are implemented as spin-locks.  Blocking
+ * on a lock will result in busying waiting.
+ *
+ * The LockTable is an array of #buckets.  Locks are "hashed" into buckets based
+ * on the lower n-1 bits of the lockId (n is such that 2^n is the number of
+ * buckets); the implementation assumes that lockIds are relatively uniformly
+ * distributed.
+ *
+ * Each bucket consists of a one or more chained \link CacheLine
+ * CacheLines\endlink, the first of which lives inline in the array of buckets.
+ * Each cache line consists of several LockTable Entry objects; each Entry
+ * contains the lockId of an acquired lock or 0; only acquired locks are
+ * kept in the LockTable.  In exception, the first entry in a bucket's first
+ * cache line is specially designated and is cast and used as the bucket's
+ * \link BucketLock BucketLock\endlink; this is a monitor-style spin-lock that
+ * protects against concurrent accesses to the same bucket.
+ *
+ * If there are two many acquired locks to fit in the bucket's first cache line,
+ * additional overflow cache lines are allocated (outside the array of buckets)
+ * and chained together.  Cache lines are never removed from buckets; lock
+ * acquisition performance in a particular bucket degrades as more cache lines
+ * are added.
+ *
+ * For best performance, the number of buckets should be set large enough so
+ * that overflow cache lines are almost never needed but small enough that the
+ * entire structure might fit in CPU cache.
```

Comment:
The above documentation is good, and it is all relevant. My only question is whether it makes sense to put it centrally here, or distributed elsewhere, nearer to the code that it describes. On the one hand, it's nice to have it all central, but on the other hand, this increases the probability that it won't get  updated when the code changes.